File: common/src/main/java/com/hartwig/actin/molecular/sort/DisruptionComparator.java
package com.hartwig.actin.molecular.sort;

import java.util.Comparator;

import com.hartwig.actin.molecular.datamodel.driver.Disruption;

import org.jetbrains.annotations.NotNull;

public class DisruptionComparator implements Comparator<Disruption> {

    @Override
    public int compare(@NotNull Disruption disruption1, @NotNull  Disruption disruption2) {
        int homozygousCompare = Boolean.compare(disruption2.isHomozygous(), disruption1.isHomozygous());
        if (homozygousCompare != 0) {
            return homozygousCompare;
        }

        int geneCompare = disruption1.gene().compareTo(disruption2.gene());
        if (geneCompare != 0) {
            return geneCompare;
        }

        return disruption1.details().compareTo(disruption2.details());
    }
}

File: datamodel/src/main/kotlin/com/hartwig/actin/datamodel/molecular/sort/driver/DisruptionComparator.kt
package com.hartwig.actin.datamodel.molecular.sort.driver

import com.hartwig.actin.datamodel.molecular.driver.Disruption

class DisruptionComparator : Comparator<Disruption> {

    private val comparator = Comparator.comparing<Disruption, Disruption>({ it }, DriverComparator())
        .thenComparing({ it }, GeneAlterationComparator())
        .thenComparing({ it.type.toString() }, String::compareTo)
        .thenComparing(Disruption::junctionCopyNumber, reverseOrder())
        .thenComparing(Disruption::undisruptedCopyNumber)
    
    override fun compare(disruption1: Disruption, disruption2: Disruption): Int {
        return comparator.compare(disruption1, disruption2)
    }
}
